javatests/arcs/android/storage/database/DatabaseImplTest.kt

```diff
@@ -2837,6 +2837,525 @@ class DatabaseImplTest {
+            ),
+            schema = PRE_UPGRADE_SCHEMA,
+            databaseVersion = 1,
+            versionMap = VersionMap("actor" to 1)
+        )
+
+        database.insertOrUpdate(backingKey1, entity1)
+        database.insertOrUpdate(collectionKey, collection)
+
+        val outputCollection = database.getCollection(collectionKey, PRE_UPGRADE_SCHEMA)
+        assertThat(outputCollection!!.values.map {
+            database.getEntity(it.reference.storageKey, PRE_UPGRADE_SCHEMA)
+        }).containsExactly(entity1)
+
+        val entity2 = DatabaseData.Entity(
+            RawEntity(
+                "entity2",
+                singletons = mapOf(
+                    "text" to "one hundred".toReferencable(),
+                    "number" to 100.0.toReferencable()
+                ),
+                collections = emptyMap()
+            ),
+            POST_UPGRADE_SCHEMA,
+            FIRST_VERSION_NUMBER,
+            VERSION_MAP
+        )
+
+        val collection2 = DatabaseData.Collection(
+            values = setOf(
+                ReferenceWithVersion(
+                    Reference("entity1", backingKey1, VersionMap("entity1" to 1)),
+                    VersionMap("actor" to 1)
+                ),
+                ReferenceWithVersion(
+                    Reference("entity2", backingKey2, VersionMap("entity2" to 1)),
+                    VersionMap("actor" to 2)
+                )
+            ),
+            schema = POST_UPGRADE_SCHEMA,
+            databaseVersion = 2,
+            versionMap = VersionMap("actor" to 2)
+        )
+
+        // The existing entity is returned as if it belongs to the new schema,
+        // but with a null value for the new field "number".
+        val outputVersionOfEntity1 = DatabaseData.Entity(
+            RawEntity(
+                "entity1",
+                mapOf("text" to "forty two".toReferencable(), "number" to null),
+                emptyMap()
+            ),
+            POST_UPGRADE_SCHEMA,
+            FIRST_VERSION_NUMBER,
+            VERSION_MAP
+        )
+
+        database.insertOrUpdate(backingKey2, entity2)
+        database.insertOrUpdate(collectionKey, collection2)
+        val outputCollection2 = database.getCollection(collectionKey, POST_UPGRADE_SCHEMA)
+        assertThat(outputCollection2!!.values.map {
+            database.getEntity(it.reference.storageKey, POST_UPGRADE_SCHEMA)
+        }).containsExactly(outputVersionOfEntity1, entity2)
+    }
+
+    @Test
+    fun canUpgradeSchema_byRemovingFields() = runBlockingTest {
+        val collectionKey = DummyStorageKey("collection")
+        val backingKey1 = DummyStorageKey("backing1")
+        val backingKey2 = DummyStorageKey("backing2")
+
+        val entity1 = DatabaseData.Entity(
+            RawEntity(
+                "entity1",
+                mapOf("text" to "forty two".toReferencable(), "number" to 42.0.toReferencable()),
+                emptyMap()
+            ),
+            POST_UPGRADE_SCHEMA,
```

Comment:
consider renaming these schemas to something like SCHEMA_ONE_FIELD SCHEMA_TWO_FIELDS since post/pre do not describe what happens in this test well.